Source/dataframe.lisp

```diff
@@ -0,0 +1,245 @@
+                                 (let ((new-row (make-array col-num :adjustable t)))
+                                   (copy-row row new-row)
+                                   (setf (elt new-rows index) new-row)))))))))
+
+(defgeneric map-data-frame (<data-frame> %row-slice% %col-slice% function)
+  (:method ((df <raw-data-frame>) row-slice col-slice fun)
+    (map-data-frame-rows
+     df row-slice
+     (lambda (row-index row)
+       (map-data-frame-cols
+        df row col-slice
+        (lambda (col-index col-name value)
+          (funcall fun row-index row col-index col-name value)))))))
+
+(defgeneric map-data-frame-rows (<data-frame> %row-slice% function)
+  (:method ((df <raw-data-frame>) row-slice fun)
+    (assert (typep row-slice '%row-slice%) nil '<row-invalid-slice>)
+    (let ((rows (rows df)))
+      (etypecase row-slice
+        ((eql t)
+         (loop for index from 0
+               for row across rows
+               do (funcall fun index row)))
+        ((cons (eql t) %row-index%)
+         (loop for index from 0 upto (cdr row-slice)
+               for row across rows
+               do (funcall fun index row)))
+        ((cons %row-index% (eql t))
+         (loop for index from (car row-slice)
+               for row = (elt rows index)
+               do (funcall fun index row)))
+        ((cons %row-index% %row-index%)
+         (loop for index from (car row-slice) upto (cdr row-slice)
+               for row = (elt rows index)
+               do (funcall fun index row)))
+        ((vector %row-index%)
+         (loop for index across row-slice
+               for row = (elt rows index)
+               do (funcall fun index row)))))))
+
+(defgeneric map-data-frame-cols (<data-frame> %row-index% %col-slice% function)
+  (:method ((df <raw-data-frame>) row col-slice fun)
+    (assert (typep row '%row-spec%)        nil '<row-invalid-index>)
+    (assert (typep col-slice '%col-slice%) nil '<col-invalid-slice>)
+    (typecase row
+      (integer (setf row (elt (rows df) row)))
+      ;; Undocumented hack to map over col names.
+      (null (setf row (make-array (length (rows df))))))
+    (let ((cols (cols df)))
+      (flet ((norm-index (col)
+               (if (integerp col)
+                   col
+                   (position col cols :test #'string=))))
+        (etypecase col-slice
+          ((eql t)
+           (loop for index from 0
+                 for name across cols
+                 for value across row
+                 do (funcall fun index name value)))
+          ((cons (eql t) %col-index%)
+           (loop with i2 = (norm-index (cdr col-slice))
+                 for index from 0 upto i2
+                 for name across cols
+                 for value across row
+                 do (funcall fun index name value)))
+          ((cons %col-index% (eql t))
+           (loop with i1 = (norm-index (car col-slice))
+                 for index from i1
```

Comment:
should be
`                 for index from i1 below (length (cols df))
`

Reply:
fixed